src/main/java/springee/doctor/DoctorController.java

```diff
@@ -0,0 +1,110 @@
+package springee.doctor;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+import javax.print.Doc;
+import java.net.URI;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+@RestController
+public class DoctorController {
+
+    private Map<Integer, Doctor> doctorMap = new HashMap<Integer, Doctor>() {{
+        put(0, new Doctor(0, "Robert Martin", "psychologist"));
+        put(1, new Doctor(1, "Josh Long", "surgeon"));
+        put(2, new Doctor(2, "Joshua Bloch", "ophthalmologist"));
+    }};
+
+    @GetMapping(value = "/doctors")
+    public List<Doctor> getDoctors (@RequestParam Optional<String> specialization,
+                                    @RequestParam Optional<String> name) {
+
+        Predicate<Doctor> specFilter = specialization.map(this::filterBySpec)
+            .orElse(doctor -> true);
+
+        Predicate<Doctor> nameFilter = name.map(this::filterByFirstLetterOfName)
+            .orElse(pet -> true);
+
+        Predicate<Doctor> complexFilter = nameFilter.and(specFilter);
+
+        return doctorMap.values().stream()
+            .filter(complexFilter)
+            .collect(Collectors.toList());
+    }
+
+    @GetMapping(value = "/doctors/{id}")
+    public ResponseEntity<?> getDoctorById(@PathVariable Integer id) {
+
+        if (doctorMap.containsKey(id)) {
+            return ResponseEntity.ok(doctorMap.get(id));
+        }
+
+        return ResponseEntity.notFound().build();
+    }
+
+    @PostMapping("/doctors")
+    public ResponseEntity<Void> createDoctor(@RequestBody Doctor doctor) {
+
+        if (doctor.getId() == null) {
+            synchronized (DoctorController.class) {
+                Integer index = doctorMap.size();
```

Comment:
this solution works but
Use `class` as monitor is not good idea. 
Better to usage: with `this` as monitor (it is default monitor for `synchronized { ... }`)
Even more better: avoid synchronization using non-blocking data types (Atomic....)
And the best solution: avoid any state and any counters